### ISSUE TITLE
Android cannot find game files

### ISSUE THREAD

**Issue by gordon-fish** [NY00123/refkeen#20]
I wanted to try out the Android version (``reflectionhle-android-20221014.apk``) of ReflectionHLE after trying ``reflectionhle-linux-x86_64-20221014.tar.xz`` on my desktop. I copied the the same data that worked on my desktop to ``Games/KDREAMS`` in the internal shared storage of my phone.

Then I loaded the app and chose ``Select game``, which presents me with this list (see screenshot):
![Screenshot_20240709-154924182](https://github.com/ReflectionHLE/ReflectionHLE/assets/19782542/eaca9722-d111-4dd0-9365-0a8aa9952cbf)

So I chose ``Add missing game version`` and then ``Primary storage``, then ``Games``, and finally ``KDREAMS`` before selecting ``Confirm``.

Then it processes for about a second before giving me a screen with the title ``No new compatible game found`` (see screenshot):
![Screenshot_20240709-155635488](https://github.com/ReflectionHLE/ReflectionHLE/assets/19782542/1d882932-675b-4aac-a0e6-75bbf12e9b3d)

It seems that it's unable to recognize any of the game files even though it can browse to the folder just fine.

Is there something that I am missing? Is there a specific location that I should be placing the game files?

**Comment by NY00123** [comment]
Hi,

Thanks for the reminder and for giving ReflectionHLE a try! I did experience the problem after changing device and briefly trying the application on it last year. As suspected by me (thanks to Discord messages from Beloko Games' founder that I found), it's due to Android 11's introduction of the scoped storage model. This model does not take just the files' locations into account, but also the types of contents.

As in your case, ReflectionHLE could be used to browse directories outside of `<storage>/Android/data/com.reflectionhle` (at least with the usual permissions granted), but it couldn't locate the files. In practice, I realized that it could access only specific kinds of files, like pictures as recognized by the system. Given the definitions of granted permissions, that makes sense.

Unfortunately, while ReflectionHLE itself may access game files under `<storage>/Android/data/com.reflectionhle/files/user_gameinsts`, other apps cannot write in the `com.reflectionhle` folder. I tried a few approaches, including connection of the device to a PC and usage of `adb shell`, and scoped storage seems to work as intended.

It might be possible to read files from the `com.reflectionhle` via a PC or another app with the right permissions, but no write to it. Additionally, while it was possible for me to delete/rename the `com.reflectionhle` folder and create a new one (possibly modified from the original), the new contents would have a different user id internally, hence being inaccessible by ReflectionHLE.

To summarize, there's no simple solution for now that I'm aware of, outside of the following ones:
1. If there's root access, make (careful) use of it.
2. Use a device with Android 10 or earlier.
3. Add the MANAGE_EXTERNAL_STORAGE to AndroidManifest.xml in the ReflectionHLE repository and make your own APK file from sources. Other code (SDL2) will also have to be built. Of course, this will require a development environment. While I can technically make an APK, I might want to wait a bit with the decision. ReflectionHLE itself still won't ask you to grant permission afterwards, but you might find a toggle somewhere in the system's settings. At least for me, it's here: Settings -> Apps -> Special app access -> All files access -> ReflectionHLE

**Comment by NY00123** [comment]
ok, you may try something now. I've attached an APK file (repackaged in a ZIP) for testing. I'll leave this issue open so it may be found by other people having the same problem. It's not impossible there won't be much more changes than what's covered here.

In addition to making use of the permission MANAGE_EXTERNAL_STORAGE instead of READ_EXTERNAL_STORAGE on Android 11 and later, it should now ask you to enable the permission if not already the case; Or at least, it should try asking that.

While this permission is usually reserved for quite specific apps to be published, I've went with manually downloadable APK files so far, so hopefully, that should work.

Changes have been based on examples from the following article, albeit eventually not deviating from the older code by much (short of refactoring an optional dialog into a separate class): <https://medium.com/@kezzieleo/manage-external-storage-permission-android-studio-java-9c3554cf79a7>

[reflectionhle-debug-12-gdadac9e.apk.zip](https://github.com/user-attachments/files/16203609/reflectionhle-debug-12-gdadac9e.apk.zip)

**Comment by gordon-fish** [comment]
Thanks, I'll give [reflectionhle-debug-12-gdadac9e.apk.zip](https://github.com/user-attachments/files/16203609/reflectionhle-debug-12-gdadac9e.apk.zip) that you linked a try and will report back.

**Comment by gordon-fish** [comment]
I installed reflectionhle-debug-12-gdadac9e.apk.zip and it was indeed able to find them all - KDREAMS, CATACOMB 3-D/ABYSS/ARMAGEDDON/APOCALYPSE, WOLF3D/SPEAR/NOAH3D. Thanks.

**Comment by NY00123** [comment]
Thanks for testing and for your original report. The issue should be fixed in the last release of ReflectionHLE, at least with compatible versions of Android.
